### PR TITLE
Abrandoned/redis async

### PR DESCRIPTION
**Pull request by abrandoned**
We have a few problems with the in memory adapter:

1. if the process dies or is killed then all memory messages are gone
2. if there is a backup on the rabbit side then it will eventually (by configuration) stop the process from pushing anything (and a memory backup can happen with bursts)

We started the idea of the redis queue to run an individual redis instance to capture all of the events and make the write durable, so we have durability and recovery combined with the ability to use disk to backup for millions of writes before we start stopping or slowing flow

@film42 @quixoten

### CODE REVIEW

**Comment by abrandoned** [comment]
made a design decision that I am less proud of (because I don't like complexity)

wondering if we should switch to Lists instead of Sets in redis  ... a set requires that we append a UUID to the message instance (which isn't bad and prevents us from somehow duplicating message pubs) but may also be costly for little gain

redis doesn't have the ability to do a multi-item `pop` on lists, this is only baked into the Set type which allows us to fetch large blocks of messages for publishing

I decided to use the set and the uuid because we do have times of large backups and being able to do multi-object `pops` seems like a better alternative than many round trips

**Comment by abrandoned** [comment]
I'll post the benchmarking code as well when we figure out where we want it for the long haul, but here are the results of the async queue vs the redis publisher (the redis publisher is approx 5,000/second with TCP and 7,000/second with Unix socket)

This is much slower than the async queue (expectedly) but it is actually faster to use TCP/Unix direct vs the first iteration which introduced an `async` object from Concurrent-Ruby to make the calls async, they were 500-1000 calls/sec slower (and caused backlogs that would result in durability concerns)

```
jruby-9.1.12.0 :001 > require "./bench"
I, [2018-01-07T14:41:23.039439 #39459]  INFO -- : Starting in-memory publisher adapter
I, [2018-01-07T14:41:23.112432 #39459]  INFO -- : Starting redis publisher adapter
Warming up --------------------------------------
         async queue     4.335k i/100ms
         redis async   182.000  i/100ms
Calculating -------------------------------------
         async queue    129.310k (±13.0%) i/s -      1.266M in  10.013416s
         redis async      6.869k (±18.8%) i/s -     64.792k in  10.014775s
 => true
```

**Comment by abrandoned** [comment]
decided to add pipelining and a memory queue to flush to redis through a pipeline and double the throughput (again, probably not necessary but an idea that hit me)

```
jruby-9.1.12.0 :001 > require "./bench"
I, [2018-01-07T22:23:22.602809 #44634]  INFO -- : Starting in-memory publisher adapter
I, [2018-01-07T22:23:22.680327 #44634]  INFO -- : Starting redis publisher adapter
Warming up --------------------------------------
         async queue     3.984k i/100ms
         redis async   493.000  i/100ms
Calculating -------------------------------------
         async queue    132.691k (±11.0%) i/s -      1.303M in   9.990988s
         redis async     15.440k (±21.2%) i/s -    145.435k in  10.025490s
 => true
```

**Review by liveh2o — APPROVED** [review]
Looks good. I think the choice to use a Set is fine. The performance tradeoffs seem sound to me.

**Comment by abrandoned** [comment]
based on discussion with @film42 moved to using a list and pipelining the pops on the consumer side .... getting more than double the throughput and will be ordered

```
jruby-9.1.12.0 :001 > require "./bench"
I, [2018-01-08T09:57:30.641041 #46114]  INFO -- : Starting in-memory publisher adapter
I, [2018-01-08T09:57:30.711206 #46114]  INFO -- : Starting redis publisher adapter
Warming up --------------------------------------
         async queue     7.885k i/100ms
         redis async     1.500k i/100ms
Calculating -------------------------------------
         async queue    371.242k (±10.8%) i/s -      3.651M in   9.997256s
         redis async     35.368k (±25.1%) i/s -    325.500k in  10.020857s
 => true
```

File: lib/active_publisher/message.rb [hunk 1]
```diff
@@ -1,3 +1,7 @@
 module ActivePublisher
-  class Message < Struct.new(:route, :payload, :exchange_name, :options); end
+  class Message < Struct.new(:route, :payload, :exchange_name, :options)
+    def initialize(*args)
```
Comment (film42) [inline lib/active_publisher/message.rb]: Looks like we can drop this again.

File: spec/lib/active_publisher/async/redis_adapter_spec.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      expect_any_instance_of(::Redis).to receive(:lpush)
+      subject.publish(route, payload, exchange_name, options)
+    end
+  end
+
+  describe "#shutdown!" do
+    # This is called when the rspec finishes. I'm sure we can make this a better test.
+  end
+
+  describe "::Message" do
```
Comment (film42) [inline spec/lib/active_publisher/async/redis_adapter_spec.rb]: We can probably remove this block since it doesn't make a ton of sense to test.

**Review by film42 — CHANGES_REQUESTED** [review]
This is shaping up to look really nice! We're for sure going to need to add tests around the new adapter and redis backed multi-op queue. If you want to pre-release as-is, that's fine, but we need more tests before a stable release.

File: lib/active_publisher/async/in_memory_adapter/consumer_thread.rb [hunk 3]
```diff
@@ -101,7 +101,7 @@ def start_thread
                 raise unknown_error
               ensure
                 # Always requeue anything that gets stuck.
-                queue.concat(current_messages) unless current_messages.nil?
+                queue.concat(current_messages) if current_messages && !current_messages.empty?
```
Comment (film42) [inline lib/active_publisher/async/in_memory_adapter/consumer_thread.rb:104]: 👍

**Review by film42 — APPROVED** [review]

**Review by film42 — APPROVED** [review]
This looks great and has been running smooth. Since this is a new feature, I think you're good to release when ready.

**Pull request merged**